### PR TITLE
Add a JVM backend debug information test, this commit is for verifying line numbers for stepping.

### PR DESCRIPTION
**Pull request by neetopia**
Running JVM instance and read stepping events from it to verify with the
test data.

### CODE REVIEW

**Comment by neetopia** [comment]
This is still a draft, just want to bring up the event logging logic to see if everyone agrees on this approach.

File: compiler/testData/debug/stepping/IfTrueThenFalse.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+fun cond() = false
+
+fun box() {
+    if (cond())
+        cond()
+    else
+         false
+}
+
+fun main(args: Array<String>) {
```
Comment (punzki) [inline compiler/testData/debug/stepping/IfTrueThenFalse.kt]: Idea: Similar to `LowLevelDebuggerTestBase`, maybe you can generate and write your own "main class" that simply calls `box()`? This way we can easily re-use existing box tests to create stepping tests.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+            classesDir.absolutePath,
+            PathUtil.getJarPathForClass(Delegates::class.java) // Add Kotlin runtime JAR
+        )
+
+        val command = arrayOf(
+            findJavaExecutable().absolutePath,
+            "-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=$DEBUG_PORT",
+            "-ea",
+            "-classpath", classpath.joinToString(File.pathSeparator),
+            classesToLoad
```
Comment (punzki) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: This won't actually work if you have multi-file tests. You can't pass `Class1,Class2` to the Java executable. For now, you'll need to assert that `classFileFactory.getClassFiles()` just returns one class. Or if you take the idea of generating a "main class" then you just pass the main class here.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 3]
```diff
@@ -0,0 +1,202 @@
+            /*
```
Comment (ting-yuan) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Remove leading white spaces

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    private fun startDebuggeeProcess(classesDir: File): Process {
+        val classesToLoad = this.classFileFactory.getClassFiles()
+            .map { it.qualifiedName }
+            .joinToString(",")
+
+        val classpath = listOf(
+            classesDir.absolutePath,
+            PathUtil.getJarPathForClass(Delegates::class.java) // Add Kotlin runtime JAR
```
Comment (punzki) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Is this needed currently for the tests?

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt (hunk 3 above)
Comment (punzki) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Nit: Formatting is off

File: compiler/testData/debug/stepping/IfTrueThenFalse.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    box()
+}
+
+// 4 1 4 7 8
```
Comment (punzki) [inline compiler/testData/debug/stepping/IfTrueThenFalse.kt]: Consider also recording the method name and maybe we can do something like:
```
// box(): 4
// cond(): 1
// box(): 4 7 8
```

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        private const val DEBUG_ADDRESS = "127.0.0.1"
+        private const val DEBUG_PORT = 5225
+    }
+
+    override fun doMultiFileTest(wholeFile: File, files: List<TestFile>) {
+        createEnvironmentWithMockJdkAndIdeaAnnotations(ConfigurationKind.ALL, *listOfNotNull(writeJavaFiles(files)).toTypedArray())
+
+        val lineNumbers = wholeFile
+            .readLines()
+            .filter { it.matches("// \\d+.*".toRegex()) }
```
Comment (ting-yuan) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Can we have a keyword here to make it less magical?

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+        try {
+            val classesDir = File(tempDirForTest, "classes").apply {
+                for (classFile in classFileFactory.getClassFiles()) {
+                    File(this, classFile.relativePath).mkdirAndWriteBytes(classFile.asByteArray())
+                }
+            }
+
+            val process = startDebuggeeProcess(classesDir)
+            waitUntil { isPortOpen() }
```
Comment (ting-yuan) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: This is inherently unreliable and will block the test frequently. Can you find a random port to use? Also, this is probably not needed if you specify a timeout in `SocketAttachingConnector.attach`.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+        return ProcessBuilder(*command).inheritIO().start()
+    }
+
+    private fun attachDebugger(): VirtualMachine {
+        val connector = SocketAttachingConnector()
+        return connector.attach(connector.defaultArguments().toMutableMap().apply {
+            getValue("port").setValue("$DEBUG_PORT")
+            getValue("hostname").setValue(DEBUG_ADDRESS)
+        })
```
Comment (ting-yuan) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Probably need a timeout and some error reporting.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        virtualMachine: VirtualMachine,
+        factory: OriginCollectingClassBuilderFactory,
+        classFileFactory: ClassFileFactory,
+        state: GenerationState
+    ) {
+        val manager = virtualMachine.eventRequestManager()
+        val loggedEvents = ArrayList<Event>()
+
+        var inBox = false
+        var finish = false
```
[the remark this answers is not in the document]
Comment (neetopia) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: The reason for using flag is that, when exiting a method, it is not guaranteed that step event will come before exit event, so if we do break, we might miss the step event in same EventSet.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+                        if (inBox) {
+                            loggedEvents.add(event)
+                        }
+                    }
+                    is MethodExitEvent -> {
+                        if (event.location().method().name() == "box") {
+                            finish = true
+                        }
+                    }
+                    else -> {
```
[the remark this answers is not in the document]
Comment (neetopia) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: I will probably add exception here notifying uncaptured events.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (punzki) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Discussed offline: According to `EventSet` documentation (https://docs.oracle.com/javase/7/docs/jdk/api/jpda/jdi/com/sun/jdi/event/EventSet.html), the only events that are grouped should be:
- BreakpointEvent
- StepEvent
- MethodEntryEvent

so what you described _should_ not happen.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+            val eventSet = virtualMachine.eventQueue().remove()
+            for (event in eventSet) {
+                when (event) {
+                    is VMDeathEvent, is VMDisconnectEvent -> {
+                        finish = true
+                    }
+                    is VMStartEvent -> {
+                        val classReq = manager.createClassPrepareRequest()
+                        classFileFactory.getClassFiles().map { classReq.addClassFilter(it.internalName) }
+                        classReq.setSuspendPolicy(SUSPEND_ALL)
```
Comment (punzki) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Discussed offline: Should be faster if you use `SUSPEND_NONE` for all these requests, and use `createVMDeathRequest()` (https://docs.oracle.com/javase/7/docs/jdk/api/jpda/jdi/com/sun/jdi/request/EventRequestManager.html#createVMDeathRequest()) to suspend before VM dies.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt (hunk 7 above)
Comment (neetopia) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Can't find a way to get a port and reserve it for later use, but definitely a valid point. Plan to address it in future.

File: compiler/testData/debug/stepping/IfTrueThenFalse.kt (hunk 5 above)
Comment (neetopia) [reply compiler/testData/debug/stepping/IfTrueThenFalse.kt]: Currently didn't see benefit of adding the function names in test data, @yanex any thoughts?

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt (hunk 4 above)
Comment (neetopia) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Mainly for extending to multi-file test cases.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+        vmLoop@ while (true) {
+            val eventSet = virtualMachine.eventQueue().remove()
+            for (event in eventSet) {
+                when (event) {
+                    is VMDeathEvent, is VMDisconnectEvent -> {
+                        finish = true
+                    }
+                    is VMStartEvent -> {
+                        val classReq = manager.createClassPrepareRequest()
+                        classFileFactory.getClassFiles().map { classReq.addClassFilter(it.internalName) }
```
[the remark this answers is not in the document]
Comment (neetopia) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Yes, for now we only care about test files.

File: compiler/testData/debug/stepping/IfTrueThenFalse.kt (hunk 1 above)
Comment (neetopia) [reply compiler/testData/debug/stepping/IfTrueThenFalse.kt]: Was try to fix this, but getting some issue with class loader not found exception, for now I will just put a //FILE: main.kt for class to be load, and put MainKt in command line arguments, will address it once I figure out the issue.

File: compiler/testData/debug/stepping/IfTrueThenFalse.kt (hunk 5 above)
Comment (neetopia) [reply compiler/testData/debug/stepping/IfTrueThenFalse.kt]: Actually I found it might be helpful, added method information for multi-file use case

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+            }
+        }
+        val actualLineNumbers = loggedEvents
+            .map { event ->
+                "${(event as LocatableEvent).location().method()}:${event.location().lineNumber()}"
+            }
+        TestCase.assertEquals(expectedLineNumbers, actualLineNumbers.joinToString(" "))
+    }
+
+    private fun isPortOpen(): Boolean {
```
Comment (madsager) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: You should be able to avoid this if you read stdout from the jvm for the port number.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/ir/AbstractIrSteppingTest.kt [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+ */
+
+package org.jetbrains.kotlin.codegen.ir
+
+import org.jetbrains.kotlin.codegen.AbstractSteppingTest
+import org.jetbrains.kotlin.config.CompilerConfiguration
+import org.jetbrains.kotlin.config.JVMConfigurationKeys
+
+abstract class AbstractIrSteppingTest : AbstractSteppingTest() {
+    override val DEBUG_PORT = 5227
```
Comment (madsager) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/ir/AbstractIrSteppingTest.kt]: Remove.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+import java.lang.IllegalStateException
+import java.lang.reflect.Method
+import java.net.Socket
+import java.nio.charset.Charset
+import java.nio.file.Files
+import kotlin.properties.Delegates
+
+abstract class AbstractSteppingTest : CodegenTestCase() {
+
+    protected open val DEBUG_PORT = 5225
```
Comment (madsager) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Remove, see later comments.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+
+    private fun startDebuggeeProcess(classesDir: File): Process {
+        val classpath = listOf(
+            classesDir.absolutePath,
+            PathUtil.getJarPathForClass(Delegates::class.java) // Add Kotlin runtime JAR
+        )
+
+        val command = arrayOf(
+            findJavaExecutable().absolutePath,
+            "-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=$DEBUG_PORT",
```
[the remark this answers is not in the document]
Comment (neetopia) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Thanks for this suggestion, addressed.

**Comment by neetopia** [comment]
@max-kammerer Hi Mike, can you take a look at this PR when you got a chance? I want to push it sooner so I can work on more debug information tests, and also pretty interested in analyzing and fixing existing failing tests

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+    data class RunJVMResult(val port: Int, val process: Process)
+
+    override fun doMultiFileTest(wholeFile: File, files: List<TestFile>) {
+        createEnvironmentWithMockJdkAndIdeaAnnotations(ConfigurationKind.ALL, *listOfNotNull(writeJavaFiles(files)).toTypedArray())
+
+        val lineNumbers = wholeFile
+            .readLines()
+            .filter { it.matches("// LINENUMBER: .*".toRegex()) }
+            .map { line ->
+                line.drop(14)
```
Comment (max-kammerer) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Please expand magic number to smth meaningful, e.g "// LINENUMBER:".length

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+                    .trim()
+            }.joinToString(" ")
+
+        loadMultiFiles(files)
+        val classBuilderFactory = OriginCollectingClassBuilderFactory(ClassBuilderMode.FULL)
+        val generationState = GenerationUtils.compileFiles(myFiles.psiFiles, myEnvironment, classBuilderFactory)
+        classFileFactory = generationState.factory
+
+        val tempDirForTest = Files.createTempDirectory("debuggerTest").toFile()
```
Comment (max-kammerer) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: KotlinTestUtils.tmpDir("debuggerTest")

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+
+        try {
+            val classesDir = File(tempDirForTest, "classes").apply {
+                for (classFile in classFileFactory.getClassFiles()) {
```
Comment (max-kammerer) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Use 'classFileFactory.writeAll(...)'

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+        val command = arrayOf(
+            findJavaExecutable().absolutePath,
+            "-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=127.0.0.1:0",
+            "-ea",
+            "-classpath", classpath.joinToString(File.pathSeparator),
+            MAIN_CLASS
+        )
+
+        val process = ProcessBuilder(*command).start()
+        val port = process.inputStream.bufferedReader().readLine()
```
Comment (max-kammerer) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Whould it work properly during parallel test execution? Would it be another port for second process?

File: compiler/testData/debug/stepping/IfTrueThenFalse.kt (hunk 1 above)
Comment (neetopia) [reply compiler/testData/debug/stepping/IfTrueThenFalse.kt]: @max-kammerer Mike, do you prefer to include class name and function name in the line number verification data or just line number only?

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+        classFileFactory = generationState.factory
+
+        val tempDirForTest = KotlinTestUtils.tmpDir("debuggerTest")
+
+        try {
+            val classesDir = File(tempDirForTest, "classes").apply {
+                classFileFactory.writeAllTo(this)
+            }
+
+            val (port, process) = startDebuggeeProcess(classesDir)
```
Comment (max-kammerer) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Is it possible to avoid new process instantiation for each test? Maybe implement some simple server?
We have smth similar in TestProcessServer.kt to run test under JDK 1.6 (compiler requires JDK 1.8).
Actuall configuration is performed in compiler/tests-different-jdk/build.gradle.kts `codegenTest(target = 6, jvm = 6, jdk = "JDK_18")`

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt (hunk 19 above)
Comment (max-kammerer) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Yes

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt (hunk 21 above)
Comment (neetopia) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Will that conflict with parallel test as paralleled tests should have different port numbers?

Comment (max-kammerer) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Actually `codegenTest(target = 6, jvm = 6, jdk = "JDK_18")` tests are not parallelized. And seems  parallelization would not work properly with current approach of configuring and shuting down server with gradle. Seems it's best to instantiate server in test VM: in such case custom test runner would be required.

Comment (neetopia) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: I see, can you give me a pointer to how to customize test runner for this purpose?

Comment (max-kammerer) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: We have several custom runners that performs system preconfiguration in clinit, e.g. JUnit3WithIdeaConfigurationRunner. You can make similar approach starting server in clinit but's it's unclear how stop server cause shutdown hook wouldn't work as same JVM could be used for another tests. 
Maybe best to switch this tests to junit4 infrastructure with @BeforeClass @AfterClass routines.

Comment (max-kammerer) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: NB: seems problem with shutdown could be solved by RunListener.testRunFinished

Comment (neetopia) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt]: Sorry it took quiet a while for updating this PR, was packed by other stuff.

I've updated this test to make it only running process creation once during setup, let me know if the approach works. I modified TestProcessServer to meet my use case.

File: compiler/testData/debug/stepping/conjunction.kt [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+
+fun getA() = true
+
+fun getB() = true
+
+fun getC() = false
+
+fun getD() = true
+
+// LINENUMBER: TestKt.box():3 TestKt.getA():13 TestKt.box():3 TestKt.box():4 TestKt.getB():15 TestKt.box():4 TestKt.box():5
```
Comment (max-kammerer) [inline compiler/testData/debug/stepping/conjunction.kt]: Sorry for delay.
I would prefer to have tests in next format:
// LINENUMBERS
// TestKt.box():3 
// TestKt.getA():13
// TestKt.box():3 
// TestKt.box():4 
// TestKt.getB():15 
// TestKt.box():4 
// TestKt.box():5
...

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/debugInformation/TestProcessServer.kt [hunk 23]
```diff
@@ -0,0 +1,109 @@
+/*
```
Comment (max-kammerer) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/debugInformation/TestProcessServer.kt]: Why not merge this stuff with one in tests-common-jvm6?

Comment (neetopia) [reply compiler/tests-common/tests/org/jetbrains/kotlin/codegen/debugInformation/TestProcessServer.kt]: Merged into existing TestProcessServer with additional logic for handling output and port allocation.

File: compiler/testData/debug/stepping/conjunction.kt (hunk 22 above)
Comment (neetopia) [reply compiler/testData/debug/stepping/conjunction.kt]: Done!

File: compiler/tests-common-jvm6/tests/org/jetbrains/kotlin/test/clientserver/TestProcessServer.kt [hunk 24]
```diff
@@ -49,6 +49,8 @@ fun getBoxMethodOrNull(aClass: Class<*>): Method? {
 //Use only JDK 1.6 compatible api
 object TestProcessServer {
 
+    val DEBUG_TEST = "--debugTest"
```
Comment (max-kammerer) [inline compiler/tests-common-jvm6/tests/org/jetbrains/kotlin/test/clientserver/TestProcessServer.kt]: 'private const val'

Comment (neetopia) [reply compiler/tests-common-jvm6/tests/org/jetbrains/kotlin/test/clientserver/TestProcessServer.kt]: I made it public to make it accessible from test class, maybe extract it somewhere?

File: compiler/tests-common-jvm6/tests/org/jetbrains/kotlin/test/clientserver/TestProcessServer.kt [hunk 25]
```diff
@@ -63,49 +65,69 @@ object TestProcessServer {
...
-        val portNumber = args[0].toInt()
-        println("Starting server on port $portNumber...")
+        if (args[0] == DEBUG_TEST) {
+            suppressOutput = true
+            allocatePort = true
+        }
 
-        val serverSocket = ServerSocket(portNumber)
-        sheduleShutdownProcess()
+        val serverSocket = if (allocatePort) {
```
Comment (max-kammerer) [inline compiler/tests-common-jvm6/tests/org/jetbrains/kotlin/test/clientserver/TestProcessServer.kt]: Maybe  best
```
        val portNumber = if (allocatePort) 0 else args[0].toInt()
        println("Starting server on port $portNumber...")
        serverSocket = ServerSocket(portNumber)
        println("...server started on port $${serverSocket.localPort}")
```

File: compiler/tests-common-jvm6/tests/org/jetbrains/kotlin/test/clientserver/TestProcessServer.kt [hunk 26]
```diff
@@ -63,49 +65,69 @@ object TestProcessServer {
...
+            printlnTest("Server stopped!")
         }
     }
 
-    private fun sheduleShutdownProcess() {
+    private fun scheduleShutdownProcess() {
         handler = scheduler.scheduleAtFixedRate({
             if (!isProcessingTask && (System.currentTimeMillis() - lastTime) >= 60 * 1000 /*60 sec*/) {
-                println("Stopping server...")
+                printlnTest("Stopping server...")
```
Comment (max-kammerer) [inline compiler/tests-common-jvm6/tests/org/jetbrains/kotlin/test/clientserver/TestProcessServer.kt]: As it one time loging I would prefer 'println' usage

File: compiler/tests-common-jvm6/tests/org/jetbrains/kotlin/test/clientserver/TestProcessServer.kt [hunk 27]
```diff
@@ -63,49 +65,69 @@ object TestProcessServer {
...
+                printlnTest("Socket established...")
+                executor.execute(ServerTest(clientSocket, suppressOutput))
             }
         }
         finally {
             handler.cancel(false)
             scheduler.shutdown()
             serverSocket.close()
-            println("Server stopped!")
+            printlnTest("Server stopped!")
```
Comment (max-kammerer) [inline compiler/tests-common-jvm6/tests/org/jetbrains/kotlin/test/clientserver/TestProcessServer.kt]: and same here

**Comment by max-kammerer** [comment]
@sfs Please also look at JvmTarget6OnJvm6:
It uses junit 4 routines to instantiate only one separate VM. It would be best to create only one external process in SteppingTestGenerated tests.

**Comment by neetopia** [comment]
> Please also look at JvmTarget6OnJvm6:
> It uses junit 4 routines to instantiate only one separate VM. It would be best to create only one external process in SteppingTestGenerated tests.

Looks like all tests generated by GenerateCompilerTests.kt are running with JUnit3 test runners, does it make sense to move the tests out of the existing generator, or modify the generator to make it able to generate Junit4 test cases?

**Comment by max-kammerer** [comment]
@neetopia I think it worth to support junit 4 test cases if it's not required to much work

**Comment by max-kammerer** [comment]
After that it would be possible to move BeforeClass/AfterClass stuff similar to one in JvmTarget6OnJvm6 into base class

**Comment by neetopia** [comment]
Changed test generator DSL to have an optional flag for using JUnit 4 infrastructure.

File: compiler/tests-common/tests/org/jetbrains/kotlin/codegen/debugInformation/AbstractSteppingTest.kt [hunk 28]
```diff
@@ -0,0 +1,274 @@
...
+                ?: File(javaBin, "java").also { assert(it.exists()) }
+        }
+
+        @AfterClass
+        @JvmStatic
+        fun tearDownTest() {
+            try {
+                process.destroy()
+            } finally {
+
```
Comment (max-kammerer) [inline compiler/tests-common/tests/org/jetbrains/kotlin/codegen/debugInformation/AbstractSteppingTest.kt]: Empty finally?

File: compiler/tests/org/jetbrains/kotlin/codegen/debugInformation/IrSteppingTestGenerated.java [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+import java.io.File;
+import java.util.regex.Pattern;
+
+/** This class is generated by {@link org.jetbrains.kotlin.generators.tests.TestsPackage}. DO NOT MODIFY MANUALLY */
+@SuppressWarnings("all")
+@TestMetadata("compiler/testData/debug/stepping")
+@TestDataPath("$PROJECT_ROOT")
+@RunWith(BlockJUnit4ClassRunner.class)
+public class IrSteppingTestGenerated extends AbstractIrSteppingTest {
+    private void runTest(String testDataFilePath) throws Exception {
```
Comment (max-kammerer) [inline compiler/tests/org/jetbrains/kotlin/codegen/debugInformation/IrSteppingTestGenerated.java]: Please add test in 'stepping' subfolder to be sure that nested suites will be generated/executed correctly (if not it could be postponed for separate pr)

File: compiler/tests/org/jetbrains/kotlin/codegen/debugInformation/IrSteppingTestGenerated.java [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+        KotlinTestUtils.runTest(this::doTest, TargetBackend.JVM_IR, testDataFilePath);
+    }
+
+    public void testAllFilesPresentInStepping() throws Exception {
```
Comment (max-kammerer) [inline compiler/tests/org/jetbrains/kotlin/codegen/debugInformation/IrSteppingTestGenerated.java]: Should be also marked with @Test

File: compiler/tests/org/jetbrains/kotlin/codegen/debugInformation/IrSteppingTestGenerated.java (hunk 29 above)
Comment (neetopia) [reply compiler/tests/org/jetbrains/kotlin/codegen/debugInformation/IrSteppingTestGenerated.java]: Can you elaborate on this? Does that mean to put the abstract test suite in /debugInformation/stepping/ folder, like /debugInformation/stepping/AbstractSteppingTest, and generate to see if generator can handle that case?

Comment (max-kammerer) [reply compiler/tests/org/jetbrains/kotlin/codegen/debugInformation/IrSteppingTestGenerated.java]: Oh no, i meant just another test data file under ```testData/debug/stepping/smth/sometest.kt```
Actually a new subsuite is generated for each folder under main one

Comment (neetopia) [reply compiler/tests/org/jetbrains/kotlin/codegen/debugInformation/IrSteppingTestGenerated.java]: Verified locally, such behavior is working. But at this moment I can not think up of test cases suits for putting in a subfolder, maybe I can add such nested test suite later in the future.

**Comment by max-kammerer** [comment]
@neetopia There is some problem with execution on Windows could you take a look: https://teamcity.jetbrains.com/viewLog.html?buildId=2551397&buildTypeId=Kotlin_dev_CompilerWithTests

**Comment by neetopia** [comment]
> @neetopia There is some problem with execution on Windows could you take a look: https://teamcity.jetbrains.com/viewLog.html?buildId=2551397&buildTypeId=Kotlin_dev_CompilerWithTests

Seems to be related to Windows command line length limit, the command for invoking process is indeed pretty long, let me figure out if I can shorten it.

**Comment by neetopia** [comment]
Changed classpath in command, should work for Windows now.

**Comment by max-kammerer** [comment]
@neetopia Thanks, but there is another error on Windows: https://teamcity.jetbrains.com/viewLog.html?buildId=2556198&buildTypeId=Kotlin_dev_CompilerWithTests

**Comment by neetopia** [comment]
@max-kammerer I added some changes for supporting space in path as well as reading jar path, can you try again to run it on Windows again to see if new changes work?

**Comment by neetopia** [comment]
@max-kammerer ping?

**Comment by max-kammerer** [comment]
@neetopia Sorry for delay, merged!
Thank you!
